zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/OrchestratorZoneService.java

```diff
@@ -88,13 +171,305 @@ private ConnectionDetails buildConnectionDetails(String zoneName, IdentityZone i
+        if(id.equals(subDomain)){
+            subDomain = null;
+        }
+        return subDomain;
+    }
+
+    public void createZone(OrchestratorZoneRequest zoneRequest) throws OrchestratorZoneServiceException,
+                                                                                   ZoneAlreadyExistsException,
+                                                                                   AccessDeniedException {
+        if (!IdentityZoneHolder.isUaa()) {
+            throw new AccessDeniedException("Zones can only be created by being authenticated in the default zone.");
+        }
+        String name = zoneRequest.getName();
+        String adminClientSecret = getAdminClientSecret(zoneRequest);
+
+        checkOrchestratorZoneExists(name);
+
+        String subdomain = zoneRequest.getParameters().getSubdomain();
+        String id = UUID.randomUUID().toString();
+        subdomain = getSubDomain(subdomain, id);
+
+        String zoneSigningKey = createSigningKey(subdomain);
+
+        IdentityZone identityZone = geIdentityZone(subdomain, name, id, zoneSigningKey);
+
+        IdentityZone created = createIdentityZone(identityZone);
+
+        if(created == null){
+            throw new OrchestratorZoneServiceException("Error while create identity zone for subdomain : " + subdomain );
+        }
+
+        createDefaultIdp(created);
+
+        createUserGroups(created);
+
+        createClient(adminClientSecret, subdomain, created);
+    }
+
+    private String getAdminClientSecret(OrchestratorZoneRequest zoneRequest) throws OrchestratorZoneServiceException {
+        String adminClientSecret = zoneRequest.getParameters().getAdminClientSecret();
+        if (!StringUtils.hasText(adminClientSecret)) {
+            throw new OrchestratorZoneServiceException(
+                "The \"" + UAA_ADMIN_SECRET + "\" field cannot contain spaces or cannot be blank.");
+        }
+        return adminClientSecret;
+    }
+
+    private String getSubDomain(String subdomain, String id) throws OrchestratorZoneServiceException {
+        String customSubdomain = getCustomSubdomain(subdomain);
+        if (customSubdomain == null) {
+            subdomain = id;
+        } else {
+            subdomain = customSubdomain;
+        }
+        return subdomain;
+    }
+
+    private void checkOrchestratorZoneExists(String name) throws ZoneAlreadyExistsException {
+        IdentityZone identityZone = null;
+        try{
+            identityZone = zoneProvisioning.retrieveByName(name);
+        } catch (ZoneDoesNotExistsException e){
+            identityZone = null;
+        }
+        if(identityZone != null){
+            throw new ZoneAlreadyExistsException("Orchestrator zone already exists for name:  " + name );
+        }
+    }
+
+    private void createClient(String adminClientSecret, String subdomain, IdentityZone created)
+        throws OrchestratorZoneServiceException {
+        String zoneId = IdentityZoneHolder.get().getId();
```

Comment:
I think, this zoneId is wrong here, If we miss the below lines as available in the existing controller,

IdentityZoneHolder.set(created);
in finally block - IdentityZoneHolder.set(previous);

Reply:
yes, Done. Thanks!